### PR TITLE
mgr/progress: Global Recovery Event in ceph -s

### PR DESCRIPTION
**Pull request by kamoltat**
Only show Global Progress event in ceph -s
Right now if you want to see all the sub events
You have to do ceph progress.

Signed-off-by: Kamoltat (Junior) Sirivadhna ksirivad@redhat.com

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by kamoltat** [comment]
This is a new PR continuing from #29467

**Review by jdurgin — APPROVED** [review]
This looks good! I think we'll want to follow up with further PRs to improve the efficiency here. For large clusters, looking through all the pg states for this event (and others) by serializing to json with every update is pretty expensive.

The PGMap keeps track of num_pg_by_state internally, so if we expose python accessors to that we don't need to iterate over all pgs to get the number of active+clean pgs.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-03_09:43:24-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-04_04:38:32-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/

File: src/mon/mon_types.h [hunk 1]
```diff
@@ -641,22 +641,31 @@ inline std::ostream& operator<<(std::ostream& out, const mon_feature_t& f) {
...
     ENCODE_FINISH(bl);
   }
   void decode(ceph::buffer::list::const_iterator& p) {
-    DECODE_START(1, p);
+    DECODE_START(2, p);
     decode(message, p);
     decode(progress, p);
+    if (struct_v >= 2){
+	decode(add_to_ceph_s, p);
+    }else {
```
Comment (tchaikov) [inline src/mon/mon_types.h:658]: please add space before `else`

File: src/mon/mon_types.h [hunk 2]
```diff
@@ -641,22 +641,31 @@ inline std::ostream& operator<<(std::ostream& out, const mon_feature_t& f) {
...
+      if (message.size()) {
```
Comment (tchaikov) [inline src/mon/mon_types.h:659]:
```suggestion
      if (!message.empty()) {
```

File: src/pybind/mgr/progress/module.py [hunk 3]
```diff
@@ -133,9 +140,9 @@ class GhostEvent(Event):
     after the event is complete.
     """
 
-    def __init__(self, my_id, message, refs, started_at, finished_at=None,
+    def __init__(self, my_id, message, refs, add_to_ceph_s, started_at, finished_at=None,
                  failed=False, failure_message=None):
-        super(GhostEvent, self).__init__(message, refs, started_at)
+        super(GhostEvent, self).__init__(message, refs, add_to_ceph_s, started_at)
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:145]:
```suggestion
        super().__init__(message, refs, add_to_ceph_s, started_at)
```

File: src/pybind/mgr/progress/module.py [hunk 4]
```diff
@@ -163,13 +170,62 @@ def to_json(self):
...
+        # type: (str, List[Any], bool, int, int) -> None
+        super(GlobalRecoveryEvent, self).__init__(message, refs, add_to_ceph_s)
+        self._add_to_ceph_s = add_to_ceph_s
+        self._progress = 0.0
+        self.id = str(uuid.uuid4()) # type: str
+        self._start_epoch = start_epoch
+        self._active_clean_num = active_clean_num
+        self._refresh()
+
+    def global_event_update_progress(self, pg_dump, log):
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:196]: is `log` used anywhere?

File: src/pybind/mgr/progress/module.py [hunk 5]
```diff
@@ -163,13 +170,62 @@ def to_json(self):
...
+        # type: (Dict, Any) -> None
+        "Update progress of Global Recovery Event"
+
+        pgs = pg_dump['pg_stats']
+        total_pg_num = len(pgs)
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:201]: please define this variable close to where it is used.

**Review by tchaikov — CHANGES_REQUESTED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-03_09:43:24-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-04_04:38:32-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/

File: src/pybind/mgr/progress/module.py [hunk 6]
```diff
@@ -16,9 +16,7 @@
 import json
 
 
-ENCODING_VERSION = 2
-
-# keep a global reference to the module so we can use it from Event methods
+ENCODING_VERSIglobal reference to the module so we can use it from Event methods
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:19]: i am not sure this is a valid python statement.

Comment (kamoltat) [reply src/pybind/mgr/progress/module.py:19]: woops!

**Comment by tchaikov** [comment]
@kamoltat looks great! could you squash these changes into a single commit?

**Comment by kamoltat** [comment]
> @kamoltat looks great! could you squash these changes into a single commit?

Done!

**Review by tchaikov — CHANGES_REQUESTED** [review]
```
2020-10-15T09:06:17.364 INFO:tasks.cephfs_test_runner:test_osd_cannot_recover (tasks.mgr.test_progress.TestProgress) ... FAIL
2020-10-15T09:06:17.365 INFO:tasks.cephfs_test_runner:
2020-10-15T09:06:17.365 INFO:tasks.cephfs_test_runner:======================================================================
2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:FAIL: test_osd_cannot_recover (tasks.mgr.test_progress.TestProgress)
2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-10-14-1541/qa/tasks/mgr/test_pro
gress.py", line 264, in test_osd_cannot_recover
2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:    self.assertEqual(len(self._all_events()), osd_count - pool_size)
2020-10-15T09:06:17.367 INFO:tasks.cephfs_test_runner:AssertionError: 2 != 1
```
https://pulpito.ceph.com/kchai-2020-10-15_08:32:19-rados-wip-kefu-testing-2020-10-14-1541-distro-basic-smithi/5526866/

**Comment by kamoltat** [comment]
> ```
> 2020-10-15T09:06:17.364 INFO:tasks.cephfs_test_runner:test_osd_cannot_recover (tasks.mgr.test_progress.TestProgress) ... FAIL
> 2020-10-15T09:06:17.365 INFO:tasks.cephfs_test_runner:
> 2020-10-15T09:06:17.365 INFO:tasks.cephfs_test_runner:======================================================================
> 2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:FAIL: test_osd_cannot_recover (tasks.mgr.test_progress.TestProgress)
> 2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-10-14-1541/qa/tasks/mgr/test_pro
> gress.py", line 264, in test_osd_cannot_recover
> 2020-10-15T09:06:17.366 INFO:tasks.cephfs_test_runner:    self.assertEqual(len(self._all_events()), osd_count - pool_size)
> 2020-10-15T09:06:17.367 INFO:tasks.cephfs_test_runner:AssertionError: 2 != 1
> ```
> 
> https://pulpito.ceph.com/kchai-2020-10-15_08:32:19-rados-wip-kefu-testing-2020-10-14-1541-distro-basic-smithi/5526866/

@tchaikov Thanks, just updated the test suite

**Comment by kamoltat** [comment]
jenkins retest make check

**Comment by kamoltat** [comment]
jenkins retest this please

**Comment by kamoltat** [comment]
http://pulpito.front.sepia.ceph.com/ksirivad-2020-10-20_14:31:56-rados-wip-mgr-progress-global-event-distro-basic-smithi/

**Comment by tchaikov** [comment]
@kamoltat no need to run 400+ tests for exercising this change. 

```
--suite rados:mgr --filter 'tasks/progress' --limit 1
```

would suffice.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-21_07:01:44-rados-wip-kefu-testing-2020-10-21-1144-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-22_03:25:39-rados-wip-kefu-testing-2020-10-21-2116-distro-basic-smithi/

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2020-10-22_10:35:40-rados:mgr-wip-mgr-progress-global-event-distro-basic-smithi/
I made some changes ran some tests today and it passed. Hope this will save you some time for testing.
@tchaikov

File: qa/tasks/mgr/test_progress.py [hunk 7]
```diff
@@ -44,12 +44,51 @@ def _events_in_progress(self):
...
+        p = self._get_progress()
+        log.info(json.dumps(p, indent=2))
+        return p['completed']
+
+    def is_osd_marked_out(self, ev):
+        ev_message = ev['message']
+        if len(ev_message) < 10:
+            return False
+        else:
+            return ev_message[-10:] == "marked out"
```
Comment (tchaikov) [inline qa/tasks/mgr/test_progress.py:60]: i think this should be
```py
return ev['message'].endswith('marked out')
```

**Comment by kamoltat** [comment]
jenkins retest this please

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2020-10-23_02:26:16-rados:mgr-wip-mgr-progress-global-event-distro-basic-smithi/

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-30_02:27:32-rados-wip-kefu-testing-2020-10-29-2343-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-30_15:27:54-rados-wip-kefu-testing-2020-10-30-1822-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-31_03:12:39-rados-wip-kefu-testing-2020-10-30-1822-distro-basic-smithi/

failure tracked by

- https://tracker.ceph.com/issues/47719
- https://tracker.ceph.com/issues/48030

**Pull request merged**